File: phase2_processing/processor.js
const crypto = require('crypto');
const logger = require('../phase6_utils/logger');

class Processor {
  constructor() {
    this.criticalKeywords = ['crash', 'failed', 'error', 'stuck', 'bug', 'refund', 'issue'];
    
    this.featureKeywords = [
      'kyc', 'onboarding', 'us stocks', 'mutual fund', 'portfolio', 'charge', 'brokerage', 
      'support', 'login', 'otp', 'federal', 'bank', 'withdraw', 'deposit', 'nps', 'etf', 'sip',
      'interface', 'ui', 'ux', 'dashboard', 'verification', 'account', 'wallet'
    ];

    this.problemKeywords = [
      'slow', 'lag', 'bad', 'poor', 'worst', 'ridiculous', 'pathetic', 'useless', 
      'difficult', 'hard', 'waste', 'cheat', 'fraud', 'hidden', 'scam', 'terrible', 'annoying'
    ];

    this.requestKeywords = [
      'please add', 'should have', 'would be better', 'wish', 'want', 'improve', 'bring',
      'need', 'suggest', 'feature', 'option'
    ];

    this.lowSignalPhrases = [
      'good', 'very good', 'nice', 'best', 'ok', 'excellent', 'super', 'perfect', 'awesome', 'great',
      'good app', 'very good app', 'nice app', 'best app', 'ok app', 'good service', 'nice service',
      'issue resolved', 'problem solved', 'thank you', 'thanks', 'very helpful', 'helpful app',
      'all good', 'everything fine', 'fast download', 'wow', 'love it', 'fantastic', 'well improve',
      'best investment app', 'excellent investment app', 'excellent app', 'the best app', 'marvelous app',
      'best app ever', 'love the app', 'amazing app', 'worst app seen', 'scam app', 'fraud app',
      'this scam and fraud', 'this is scam app', 'worst app ever seen'
    ];

    // Expanded common English function words for better heuristic language detection
    this.englishStopWords = [
      'the', 'and', 'this', 'that', 'with', 'from', 'have', 'your', 'for', 'not', 'are', 'was', 'but',
      'they', 'their', 'there', 'which', 'about', 'when', 'more', 'what', 'some', 'could', 'them',
      'is', 'it', 'to', 'my', 'of', 'on', 'in', 'be', 'at', 'an', 'as', 'by', 'if', 'or', 'so'
    ];
  }

  isEnglish(text) {
    const words = text.toLowerCase().split(/\s+/);
    // Simple blacklist for very common non-English markers found in Hinglish/other languages
    const nonEnglishMarkers = ['hai', 'hi', 'hi', 'ko', 'ka', 'ke', 'ki', 'se', 'tha', 'raha'];
    if (words.some(w => nonEnglishMarkers.includes(w))) return false;

    const englishWordCount = words.filter(word => this.englishStopWords.includes(word)).length;
    
    // Require at least 2 English stop words for quality assurance in long reviews
    if (words.length > 8) return englishWordCount >= 2;
    // For short reviews, at least one solid English function word is required
    return englishWordCount > 0;
  }

  isIncoherent(text) {
    if (/(.)\1{4,}/.test(text)) return true;
    const words = text.split(/\s+/);
    for (const word of words) {
        if (word.length > 20 && !/[aeiou]/i.test(word)) return true;
    }
    return false;
  }

  isInsightful(text) {
    const lowerText = text.toLowerCase().trim();
    const wordCount = text.split(/\s+/).length;
    const hasFeature = this.featureKeywords.some(kw => lowerText.includes(kw));
    const cleanLower = lowerText.replace(/[^\w\s]/g, '').trim();

    // 1. Blacklist check early (explicitly useless phrases)
    if (this.lowSignalPhrases.some(phrase => cleanLower.includes(phrase) && cleanLower.length < phrase.length + 5)) {
        // If it's just the phrase or the phrase plus 1 small word, it's low signal
        return false;
    }

    // 2. Minimum 4 words to avoid ultra-short vague text
    if (wordCount < 4) return false;

    // 3. Stricter Detail Rule:
    // If it lacks a specific feature keyword, require more depth (min 8 words) for problems/bugs
    if (!hasFeature && wordCount < 8) return false;

    // 4. Feature-specific feedback is generally kept if it has some length
    if (hasFeature && wordCount >= 3) return true;

    // 5. Whitelists (Only if they passed the stricter length test above)
    if (this.criticalKeywords.some(kw => lowerText.includes(kw))) return true;
    if (this.problemKeywords.some(kw => lowerText.includes(kw))) return true;
    if (this.requestKeywords.some(kw => lowerText.includes(kw))) return true;

    // 6. Detailed User Experience (15+ words if no keywords)
    if (wordCount >= 15 && !this.lowSignalPhrases.includes(cleanLower)) return true;

    return false;
  }

  stripEmojis(text) {
    // This regex matches most emojis including country flags and skin tone modifiers
    return text.replace(/(\u00a9|\u00ae|[\u2000-\u3300]|\ud83c[\ud000-\udfff]|\ud83d[\ud000-\udfff]|\ud83e[\ud000-\udfff])/g, '');
  }

  normalizeForDedup(text) {
    return text.toLowerCase().trim().replace(/[^\w\s]/g, '').replace(/\s+/g, ' ');
  }

  process(reviews) {
    logger.info(`Phase 2: Processing ${reviews.length} reviews (Aggressive Filtering)...`);

    const uniqueRawMap = new Map();
    reviews.forEach(r => {
      if (!r.text) return;
      const normalized = this.normalizeForDedup(r.text);
      if (!uniqueRawMap.has(normalized)) {
          uniqueRawMap.set(normalized, r);
      }
    });

    const uniqueReviews = Array.from(uniqueRawMap.values());
    logger.info(`Phase 2: After fuzzy deduplication, ${uniqueReviews.length} unique raw reviews remain.`);

    const cleaned = uniqueReviews.filter(r => {
        const text = this.stripEmojis(r.text.trim());
        const userName = (r.userName || '').toLowerCase().trim();
        const lowerText = text.toLowerCase().trim();

        if (this.isIncoherent(text)) return false;
        if (lowerText === userName) return false;
        if (!this.isEnglish(text)) return false;
        if (!this.isInsightful(text)) return false;

        return true;
    });

    const processed = cleaned.map(r => {
        const cleanText = r.text.trim();
        const scrubbedText = cleanText
          .replace(/[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g, '[EMAIL]')
          .replace(/\b\d{10}\b/g, '[PHONE]');

        return {
          reviewId: r.id || crypto.createHash('md5').update(r.text + r.date).digest('hex'),
          text: scrubbedText,
          rating: r.score,
          date: new Date(r.date).toISOString(),
          helpfulCount: r.thumbsUp || 0
        };
    });

    logger.info(`Phase 2: Aggressive cleaning complete. ${processed.length} high-fidelity reviews retained.`);
    return processed;
  }
}

module.exports = new Processor();

File: phase2_processing/processor.py
import re
import hashlib
import datetime
from typing import List, Dict, Any

class Processor:
    def __init__(self):
        self.critical_keywords = ['crash', 'failed', 'error', 'stuck', 'bug', 'refund', 'issue']
        
        self.feature_keywords = [
            'kyc', 'onboarding', 'us stocks', 'mutual fund', 'portfolio', 'charge', 'brokerage', 
            'support', 'login', 'otp', 'federal', 'bank', 'withdraw', 'deposit', 'nps', 'etf', 'sip',
            'interface', 'ui', 'ux', 'dashboard', 'verification', 'account', 'wallet'
        ]

        self.problem_keywords = [
            'slow', 'lag', 'bad', 'poor', 'worst', 'ridiculous', 'pathetic', 'useless', 
            'difficult', 'hard', 'waste', 'cheat', 'fraud', 'hidden', 'scam', 'terrible', 'annoying'
        ]

        self.request_keywords = [
            'please add', 'should have', 'would be better', 'wish', 'want', 'improve', 'bring',
            'need', 'suggest', 'feature', 'option'
        ]

        self.low_signal_phrases = [
            'good', 'very good', 'nice', 'best', 'ok', 'excellent', 'super', 'perfect', 'awesome', 'great',
            'good app', 'very good app', 'nice app', 'best app', 'ok app', 'good service', 'nice service',
            'issue resolved', 'problem solved', 'thank you', 'thanks', 'very helpful', 'helpful app',
            'all good', 'everything fine', 'fast download', 'wow', 'love it', 'fantastic', 'well improve',
            'best investment app', 'excellent investment app', 'excellent app', 'the best app', 'marvelous app',
            'best app ever', 'love the app', 'amazing app', 'worst app seen', 'scam app', 'fraud app',
            'this scam and fraud', 'this is scam app', 'worst app ever seen'
        ]

        # Common English stop words for heuristic language detection
        self.english_stop_words = [
            'the', 'and', 'this', 'that', 'with', 'from', 'have', 'your', 'for', 'not', 'are', 'was', 'but',
            'they', 'their', 'there', 'which', 'about', 'when', 'more', 'what', 'some', 'could', 'them',
            'is', 'it', 'to', 'my', 'of', 'on', 'in', 'be', 'at', 'an', 'as', 'by', 'if', 'or', 'so'
        ]

    def is_english(self, text: str) -> bool:
        words = text.lower().split()
        if not words: return False
        
        # Simple blacklist for common non-English markers (Hinglish)
        non_english_markers = ['hai', 'hi', 'ko', 'ka', 'ke', 'ki', 'se', 'tha', 'raha']
        if any(w in words for w in non_english_markers):
            return False

        english_word_count = sum(1 for word in words if word in self.english_stop_words)
        
        if len(words) > 8:
            return english_word_count >= 2
        return english_word_count > 0

    def is_incoherent(self, text: str) -> bool:
        # Check for repetitive characters (e.g., "aaaaa")
        if re.search(r'(.)\1{4,}', text):
            return True
        
        words = text.split()
        for word in words:
            # Long words without vowels
            if len(word) > 20 and not re.search(r'[aeiou]', word, re.I):
                return True
        return False

    def is_insightful(self, text: str) -> bool:
        lower_text = text.lower().strip()
        words = lower_text.split()
        word_count = len(words)
        
        has_feature = any(kw in lower_text for kw in self.feature_keywords)
        
        # Clean alpha-numeric only for phrase checking
        clean_lower = re.sub(r'[^\w\s]', '', lower_text).strip()

        # 1. Blacklist check (explicitly useless phrases)
        if any(phrase in clean_lower and len(clean_lower) < len(phrase) + 5 for phrase in self.low_signal_phrases):
            return False

        # 2. Minimum 4 words to avoid ultra-short vague text
        if word_count < 4:
            return False

        # 3. Stricter Detail Rule: If no feature keyword, require more depth
        if not has_feature and word_count < 8:
            return False

        # 4. Feature-specific feedback is generally kept
        if has_feature and word_count >= 3:
            return True

        # 5. Whitelists
        if any(kw in lower_text for kw in self.critical_keywords): return True
        if any(kw in lower_text for kw in self.problem_keywords): return True
        if any(kw in lower_text for kw in self.request_keywords): return True

        # 6. Detailed User Experience (15+ words if no keywords)
        if word_count >= 15 and clean_lower not in self.low_signal_phrases:
            return True

        return False

    def strip_emojis(self, text: str) -> str:
        # Simple regex-based emoji stripping
        return re.sub(r'[^\x00-\x7F]+', '', text)

    def normalize_for_dedup(self, text: str) -> str:
        return re.sub(r'\s+', ' ', re.sub(r'[^\w\s]', '', text.lower().strip()))

    def process(self, reviews: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        unique_raw_map = {}
        for r in reviews:
            text = r.get('content') or r.get('text')
            if not text: continue
            
            normalized = self.normalize_for_dedup(text)
            if normalized not in unique_raw_map:
                unique_raw_map[normalized] = r

        unique_reviews = list(unique_raw_map.values())
        
        cleaned = []
        for r in unique_reviews:
            raw_text = r.get('content') or r.get('text', '')
            text = self.strip_emojis(raw_text.strip())
            user_name = (r.get('userName') or '').lower().strip()
            lower_text = text.lower().strip()

            if self.is_incoherent(text): continue
            if lower_text == user_name: continue
            if not self.is_english(text): continue
            if not self.is_insightful(text): continue

            cleaned.append(r)

        processed = []
        for r in cleaned:
            raw_text = r.get('content') or r.get('text', '').strip()
            
            # PII Scrubbing
            scrubbed_text = re.sub(r'[a-zA-Z0-9._%+-]+@ [a-zA-Z0-9.-]+\.[a-zA-Z]{2,}', '[EMAIL]', raw_text)
            scrubbed_text = re.sub(r'\b\d{10}\b', '[PHONE]', scrubbed_text)

            review_date = r.get('at') or r.get('date')
            if isinstance(review_date, str):
                try:
                    # Handle different date formats
                    if review_date.endswith('Z'):
                        date_obj = datetime.datetime.fromisoformat(review_date.replace('Z', '+00:00'))
                    else:
                        date_obj = datetime.datetime.fromisoformat(review_date)
                except:
                    date_obj = datetime.datetime.now()
            else:
                date_obj = review_date or datetime.datetime.now()

            processed.append({
                'reviewId': r.get('reviewId') or r.get('id') or hashlib.md5(f"{raw_text}{review_date}".encode()).hexdigest(),
                'text': scrubbed_text,
                'rating': r.get('score') or r.get('rating'),
                'date': date_obj.isoformat() if hasattr(date_obj, 'isoformat') else str(date_obj),
                'helpfulCount': r.get('thumbsUpCount') or r.get('thumbsUp') or 0
            })

        return processed

processor = Processor()
